src/aleph/storage.py

```diff
@@ -22,113 +24,207 @@
     """Deserialize ``s`` (a ``str``, ``bytes`` or ``bytearray`` instance
     containing a JSON document) to a Python object in an asynchronous executor."""
     return await run_in_executor(None, json.loads, s)
 
 
-async def get_message_content(message: Dict):
+async def get_message_content(message: Dict) -> JsonContent:
     item_type: str = message.get("item_type", ItemType.IPFS)
+    item_hash = message["item_hash"]
 
     if item_type in (ItemType.IPFS, ItemType.Storage):
-        return await get_json(message["item_hash"], engine=ItemType(item_type))
+        return await get_json(item_hash, engine=ItemType(item_type))
     elif item_type == ItemType.Inline:
         if "item_content" not in message:
-            LOGGER.warning(f"No item_content in message {message.get('item_hash')}")
-            return -1, 0  # never retry, bogus data
+            error_msg = f"No item_content in message {message.get('item_hash')}"
+            LOGGER.warning(error_msg)
+            raise InvalidContent(error_msg)  # never retry, bogus data
         try:
             item_content = await json_async_loads(message["item_content"])
-        except (json.JSONDecodeError, KeyError):
-            LOGGER.exception("Can't decode JSON")
-            return -1, 0  # never retry, bogus data
-        return item_content, len(message["item_content"])
+        except (json.JSONDecodeError, KeyError) as e:
+            error_msg = f"Can't decode JSON: {e}"
+            LOGGER.exception(error_msg)
+            raise InvalidContent(error_msg)
+
+        return JsonContent(
+            hash=item_hash,
+            source=ContentSource.INLINE,
+            value=item_content,
+            raw_content=message["item_content"],
+        )
     else:
-        LOGGER.exception("Unknown item type: %s", item_type)
-        return (
-            None,
-            0,
-        )  # unknown, could retry later? shouldn't have arrived this far though.
+        # unknown, could retry later? shouldn't have arrived this far though.
+        error_msg = f"Unknown item type: '{item_type}'."
+        raise ContentCurrentlyUnavailable(error_msg)
+
+
+async def fetch_content_from_network(
+    content_hash: str, engine: ItemType, timeout: int
+) -> Optional[bytes]:
+    content = None
+    enabled_clients = app["config"].p2p.clients.value
+
+    if "protocol" in enabled_clients:
+        streamer = get_streamer()
+        content = await streamer.request_hash(content_hash)
+
+    if content is None and "http" in enabled_clients:
+        content = await p2p_http_request_hash(content_hash, timeout=timeout)
+
+    if content is not None:
+        await verify_content_hash(content, engine, content_hash)
+
+    return content
+
+
+async def compute_content_hash_ipfs(
+    content: bytes, expected_hash: str
+) -> Optional[str]:
+    """
+    Computes the IPFS hash of the content.
+    :param content: Content to hash.
+    :param expected_hash: Expected hash of the content.
```

Comment:
It seems strange to have a function `compute_hash` expecting a mandatory `expected_hash`. Should it be optional ? 

Or should the functions be named `compare_content_hash_ipfs` or `validate_content_hash_ipfs` ?
I feel like this splitting between `verify_content_hash` and `compute_content_hash_ipfs` is odd, and it could have more sense to have:

```python

 async def verify_content_hash_ipfs(content: bytes, expected_hash: str):
    ...

 async def verify_content_hash_sha256(content: bytes, expected_hash: str):
    ...

 async def verify_content_hash(content: bytes, engine: ItemType, expected_hash: str):
    ...
```

Reply:
I split it this way to reduce duplication. The error handling is the same for sha256 and IPFS hashes, we just use a different function to compute them. I'm not 100% happy about the signature either but IMO it's worth it.